src/charm.py

```diff
@@ -224,18 +217,23 @@ def _generate_kiali_layer() -> Layer:
     def _get_prometheus_source_url(self):
         """Get the Prometheus source configuration.
 
+        Returns, in this order, the first of:
+        * prometheus's ingress_url
+        * prometheus's direct_url
+
         Raises a SourceNotAvailableError if there are no sources or the data is not complete.
         """
-        if not (prometheus_sources := self._prometheus_source.sources):
-            raise PrometheusSourceError("No Prometheus sources available")
-        if len(prometheus_sources) > 1:
-            raise PrometheusSourceError("Multiple Prometheus sources available, expected only one")
-        if not (url := prometheus_sources[0].get("url", None)):
-            raise PrometheusSourceError("Prometheus source data is incomplete - url not available")
-        return url
+        if not len(self._prometheus_source.relations) == 1:
+            raise PrometheusSourceError("Missing required relation to prometheus provider")
+        if not (prometheus_data := self._prometheus_source.get_data()):
+            raise PrometheusSourceError(
+                "Prometheus relation established, but data is missing or invalid"
+            )
+        # Return ingress_url if not None, else direct_url
+        return str(prometheus_data.ingress_url or prometheus_data.direct_url)
```

Comment:
Don't let me stop anyone from reviewing things :D  But this section will get refactored by other work for #28 anyway so I suggest not getting too caught up with any complaints here.  Keep them for the next review :)